### ISSUE TITLE
Clarifications on Multi-Threading and BackgroundTask on native plugin APIs

### ISSUE THREAD

**Issue by SecondNewtonLaw** [Vector35/binaryninja-api#6109]
**Is this a general concept that needs to be documented or a specific API?**
Multi-Threading in the C++ API. How can we achieve work in the background without locking the main thread (and thus the UI)? Mainly involving the usage of BackgroundTask, as I attempted to use the structure but have been met with nothing other than crashes (even despite using the function that appears to synchronize things when in parallel `BinaryNinja::RunOnMainThreadAndWait` and related functions, but still leading to a crash). This is particularly frustrating when developing plugins that take some time to do their work, i.e., signature matchers, RTTI scanners, and other work that involves, in general, some significant workload with analyzing assembly/BNIL.

**What concept or API needs more documentation?**
The usage of `BinaryNinja::BackgroundTask` to create concurrent background tasks that prevent locks in the UI and allow plugins to execute in parallel (or if not in parallel by a user control thread in the form of Binary Ninja worker), and in general terms, how should we approach multithreading on the low-level APIs (Rust, C++, etc.)?

**Is this documentation related to a specific API?**
C++ and Rust, as the Python API seems to be able to achieve feats like this, but my focus mainly lies on the C++ API, as it is the one I'm trying to develop plugins in.

**Are there any known examples of people using this API/concept?**
Whilst I suppose there have been attempts, all public C++ Binary Ninja plugins I have seen so far have never written code that does not lock the UI for a task. This is damaging in some cases, more considering that other APIs such as the Python API appear to be able to do such feat.

**Comment by emesare** [comment]
`BinaryNinja::BackgroundTask` does nothing other than add a label at the bottom alerting the user to some background task, and optionally a way to cancel a task generically. To run things off the main thread you are responsible for creating the thread and scheduling (assuming you are not using function or module level workflows)

See this example: https://github.com/Vector35/binaryninja-api/blob/dev/examples/background_task

One thing to note is that example does not actually work on the view pointer passed to the `Inspire` function, it is important that if you _do_ to take it as a ref and move it into the thread, to avoid the pointer being invalidated.

Documentation is sparse for C++ so your best bet for documentation is to look at the python api and work backwards. The above is explained in less words [here](https://dev-api.binary.ninja/binaryninja.plugin-module.html#binaryninja.plugin.BackgroundTask).

For some more examples you can see how to use module level workflows to run your scanner code at certain parts of analysis (i.e. not on the main thread, and not managed by you): https://github.com/Vector35/binaryninja-api/blob/dev/plugins/msvc_rtti/plugin.cpp and for rust api you can also see an example here: https://github.com/Vector35/binaryninja-api/blob/dev/plugins/warp/src/plugin/workflow.rs

As for how to avoid locking, this is definitely something we should document. It unfortunately, would be a massive undertaking to properly document the locking behavior of the user facing API's, if it did happen it would be limited to functions known _not_ to lock would be documented, as those tend to be accessors or otherwise less reentrant (a function that calls another function which calls a function that might lock is hard to identify).

**Comment by emesare** [comment]
Back of the napkin code to use as an example

```cpp
#include <thread>

using namespace BinaryNinja;

static Ref<BackgroundTask> testBackgroundTask = nullptr;

void Test(Ref<BinaryView> view)
{
	std::thread testThread([view = std::move(view)]() {
		testBackgroundTask = new BackgroundTask("This is a test...", false);
		// Do something with the view on a seperate thread.
		testBackgroundTask->Finish();
	});
	testThread.detach();
}

extern "C" {
	BN_DECLARE_CORE_ABI_VERSION

	BINARYNINJAPLUGIN bool CorePluginInit()
	{
		PluginCommand::Register("Test", "Run off main thread", Test);
		return true;
	}
}
```

**Comment by SecondNewtonLaw** [comment]
> `BinaryNinja::BackgroundTask` does nothing other than add a label at the bottom alerting the user to some background task, and optionally a way to cancel a task generically. To run things off the main thread you are responsible for creating the thread and scheduling (assuming you are not using function or module level workflows)
> 
> See this example: https://github.com/Vector35/binaryninja-api/blob/dev/examples/background_task
> 
> One thing to note is that example does not actually work on the view pointer passed to the `Inspire` function, it is important that if you _do_ to take it as a ref and move it into the thread, to avoid the pointer being invalidated.
> 
> Documentation is sparse for C++ so your best bet for documentation is to look at the python api and work backwards. The above is explained in less words [here](https://dev-api.binary.ninja/binaryninja.plugin-module.html#binaryninja.plugin.BackgroundTask).
> 
> For some more examples you can see how to use module level workflows to run your scanner code at certain parts of analysis (i.e. not on the main thread, and not managed by you): https://github.com/Vector35/binaryninja-api/blob/dev/plugins/msvc_rtti/plugin.cpp and for rust api you can also see an example here: https://github.com/Vector35/binaryninja-api/blob/dev/plugins/warp/src/plugin/workflow.rs
> 
> As for how to avoid locking, this is definitely something we should document. It unfortunately, would be a massive undertaking to properly document the locking behavior of the user facing API's, if it did happen it would be limited to functions known _not_ to lock would be documented, as those tend to be accessors or otherwise less reentrant (a function that calls another function which calls a function that might lock is hard to identify).

So, in order to do _proper_ multi threading we must take the BinaryNinja::BinaryView as a BinaryNinja::Ref? If so, thank you for the information, I will apply it on my little modifications I will be doing, however behaviour like this I believe should have clearer documentation, which is why I decided to make this issue. Had I known that I could use a BinaryNinja Ref I would have tried it, but the documentation was fairly fuzzy on that regard. I do understand the part of tracking locking operations being complicated, however I believe functions could get some kind of naming convention, like an Async postfix. Although they do have a AndWait post fix, some parts of the code may lock and not specify it.

Thanks for the clarification however

**Comment by xusheng6** [comment]
I do not think using `BinaryNinja::Ref` can help you avoid the crash. That is more like related to ref-counting of the binary view object and avoid memory leak. So in other words, yes, you should probably use a `BinaryNinja::Ref`, but your crash probably needs some other fixes.

I am not particularly sure about the exact cause of your crash. Is there any chance you can share your code with us, either in public or in private? Conceptually, if you wish to do some background work and not blocking the UI, your work needs to be read-only on the view, because if both the user and the background thread is modifying the binary view, it can go wrong catastrophically. A possible strategy is to do the bulk analysis without blocking the UI, but apply any necessary changes in a blocking way

**Comment by SecondNewtonLaw** [comment]
I do not think using `BinaryNinja::Ref` can help you avoid the crash. That is more like related to ref-counting of the binary view object and avoiding memory leak. So in other words, yes, you should probably use a `BinaryNinja::Ref`, but your crash probably needs some other fixes.
> 
> I am not particularly sure about the exact cause of your crash. Is there any chance you can share your code with us, either in public or in private? Conceptually, if you wish to do some background work without blocking the UI, your work needs to be read-only on the view, because if both the user and the background thread are modifying the binary view, it can go catastrophically wrong. A possible strategy is to do the bulk analysis without blocking the UI, but apply any necessary changes in a blocking way

Utilizing a Ref smart pointer solved the issue indeed; however, the problem here roots more from the fact that using a Ref doesn't make it clear that will make the object live more than it would otherwise if it weren't used. I believe that it should be specified on the documentation if it does do such behaviour. As for the writes to the View, it appears it automatically ran on the main thread, making the explicit calls not really required, as I have not suffered any crashes of the sort ever since I used the Ref smart pointer. If any issue arises, I will make use of the function to run on the main thread, and if that does not solve it, I will make another issue.

So yeah, in summary, I'd say maybe a tweak on the documentation to specify that the callback can take a Ref smart pointer can help others who want to do multi-threading.
